openfl-tutorials/interactive_api/MXNet_landmarks/envoy/landmark_shard_descriptor.py

```diff
@@ -0,0 +1,170 @@
+# Copyright (C) 2021-2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+"""Landmarks Shard Descriptor."""
+
+import json
+import shutil
+from hashlib import md5
+from logging import getLogger
+from pathlib import Path
+from random import shuffle
+from typing import Dict
+from typing import List
+from zipfile import ZipFile
+
+import numpy as np
+import pandas as pd
+from kaggle.api.kaggle_api_extended import KaggleApi
+
+from openfl.interface.interactive_api.shard_descriptor import ShardDataset
+from openfl.interface.interactive_api.shard_descriptor import ShardDescriptor
+
+logger = getLogger(__name__)
+
+
+class LandmarkShardDataset(ShardDataset):
+    """Landmark Shard dataset class."""
+
+    def __init__(self, dataset_dir: Path,
+                 rank: int = 1, worldsize: int = 1) -> None:
+        """Initialize LandmarkShardDataset."""
+        self.rank = rank
+        self.worldsize = worldsize
+        self.dataset_dir = dataset_dir
+        self.img_names = list(self.dataset_dir.glob('img_*.npy'))
+
+        # Sharding
+        self.img_names = self.img_names[self.rank - 1::self.worldsize]
+        # Shuffling the results dataset after choose half pictures of each class
+        shuffle(self.img_names)
+
+    def __getitem__(self, index) -> np.ndarray:
+        """Return a item by the index."""
+        # Get name key points file
+        # f.e. image name:  'img_123.npy, corresponding name of the key points: 'keypoints_123.npy'
+        kp_name = str(self.img_names[index]).replace('img', 'keypoints')
+        return np.load(self.img_names[index]), np.load(self.dataset_dir / kp_name)
+
+    def __len__(self) -> int:
+        """Return the len of the dataset."""
+        return len(self.img_names)
+
+
+class LandmarkShardDescriptor(ShardDescriptor):
+    """Landmark Shard descriptor class."""
+
+    def __init__(self, data_folder: str = 'data',
+                 rank_worldsize: str = '1, 1',
+                 **kwargs) -> None:
+        """Initialize LandmarkShardDescriptor."""
+        super().__init__()
+        # Settings for sharding the dataset
+        self.rank, self.worldsize = map(int, rank_worldsize.split(','))
+
+        self.data_folder = Path.cwd() / data_folder
+        self.download_data()
+
+        # Calculating data and target shapes
+        ds = self.get_dataset()
+        sample, target = ds[0]
+        self._sample_shape = [str(dim) for dim in sample.shape]
+        self._target_shape = [str(len(target.shape))]
+
+        if self._target_shape[0] != '1':
+            raise ValueError('Target has a wrong shape')
+
+    def process_data(self, name_csv_file) -> None:
```

Comment:
Can we do processing in runtime, without saving additional files?

Reply:
Yes, we can, but then we have to process csv file every time we start the experiment and keep all images in RAM. I guess processing them just once is better.

Reply:
If I understand the situation correctly, there is a single csv file with all the labels, and you read and split it in at Envoy start time.
In this situation, you do not need to save separate labels to disk, just keep them in memory.

Reply:
Yes, you are right, we can do it. I just thought that will be easier to process them the same way as a picture.